core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java

```diff
@@ -0,0 +1,231 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file.meta;
+
+import alluxio.AlluxioURI;
+import alluxio.exception.InvalidPathException;
+import alluxio.master.file.meta.options.MountInfo;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.util.ThreadFactoryUtils;
+import alluxio.util.io.PathUtils;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import io.netty.util.internal.chmv8.ConcurrentHashMapV8;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * This is a {@link UfsAbsentPathCache} which implements asynchronous addition and removal to the
+ * cache, since the processing of the path may be slow.
+ */
+@ThreadSafe
+public final class UfsAbsentPathAsyncCache implements UfsAbsentPathCache {
+  private static final Logger LOG = LoggerFactory.getLogger(UfsAbsentPathAsyncCache.class);
+  /** Number of threads for the async pool. */
+  private static final int NUM_THREADS = 50;
+  /** Number of seconds to keep threads alive. */
+  private static final int THREAD_KEEP_ALIVE_SECONDS = 60;
+  /** Number of paths to cache. */
+  private static final int MAX_PATHS = 100000;
```

Comment:
Should these values be configurable or is the impact minimal as long as the numbers are reasonable?